src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java

```diff
@@ -0,0 +1,67 @@
+package com.atlassian.db.replica.internal;
+
+import org.junit.Test;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class LazyReferenceTest {
+
+    @Test
+    public void shouldCreateValueOnce() {
+        final CountingReference countingReference = new CountingReference();
+
+        countingReference.get();
+        countingReference.get();
+        countingReference.get();
+
+        assertThat(countingReference.getCounter()).isEqualTo(1);
+    }
+
+    @Test
+    public void shouldCreateValueOnceWhileAccessedConcurrently() throws InterruptedException {
+        final CountingReference countingReference = new CountingReference();
+        final int threads = 128;
+        final ExecutorService executor = Executors.newFixedThreadPool(threads);
+        final CountDownLatch start = new CountDownLatch(threads);
+        final CountDownLatch end = new CountDownLatch(threads);
+        for (int i = 0; i < threads; i++) {
+            executor.submit(() -> fetchReference(countingReference, start, end));
+        }
+        end.await();
+
+        assertThat(countingReference.getCounter()).isEqualTo(1);
+        executor.shutdown();
+    }
+
+    private void fetchReference(CountingReference countingReference, CountDownLatch start, CountDownLatch end) {
+        start.countDown();
+        try {
+            start.await();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        countingReference.get();
+        end.countDown();
+    }
```

Comment:
We can get a cleaner synchronization code if we turn the `Runnable` into a `Callable`
```suggestion
    private int fetchReference(
        LazyReference<Integer> reference,
        CountDownLatch start,
        CountDownLatch end
    ) throws InterruptedException {
        start.countDown();
        start.await();
        int value = reference.get();
        end.countDown();
        return value;
    }
```
If we don't want to rely on `Integer`, then we can turn `void` into `Void` and we'll be good too.